src/components/memorygame/MemoryGameCards/index.jsx

```diff
@@ -0,0 +1,171 @@
+    setSelectedCards((prevState) => [...prevState, clickedCardState]);
+  };
+
+  // matched cards should have different cardKey, same image key, isFlipped:true
+  // function below tests it
+  const doesMatchLastCard = (prevCardState, lastCardState) => {
+    return (
+      prevCardState.cardKey !== lastCardState.cardKey &&
+      prevCardState.img.imgKey === lastCardState.img.imgKey &&
+      prevCardState.isFlipped &&
+      lastCardState.isFlipped
+    );
+  };
+
+  // Triggered when user clicks on a card(selectedCards change)
+  // checks selected cards to see if they are a match
+  useEffect(() => {
+    // only triggered if "NUM_REQUIRED_MATCHES" amount of cards are clicked
+    if (selectedCards.length !== NUM_REQUIRED_MATCHES) {
+      return;
+    }
+    // match condition of selectedCards will be based on card that is added last to selectedCards
+    const lastCardState = selectedCards[selectedCards.length - 1];
+    const matchesOfLastCardState = selectedCards.filter((prevCard) =>
+      doesMatchLastCard(prevCard, lastCardState)
+    );
+    // matchesOfLastCardState doesnt include last card
+    const matchedCardsArr = [...matchesOfLastCardState, lastCardState];
+    const copyOfCardStates = [...cardStates];
+    // matchedCards is an array states that holds successfully matched card states
+    if (matchedCardsArr.length === NUM_REQUIRED_MATCHES) {
+      matchedCardsArr.forEach((matchedCard) => {
+        // current index is needed to keep the order of cards displayed
+        const targetIndex = copyOfCardStates.findIndex(
+          (card) => card.cardKey === matchedCard.cardKey
+        );
+        // if isMatched is true card image gets replaced by success symbol
+        copyOfCardStates[targetIndex] = {
+          ...copyOfCardStates[targetIndex],
+          isMatched: true,
+        };
+      });
+      // to display 1 fact per match,
+      // only one of the matched cards(selectedCards[0]) should be passed as state
+      setMatchedCards((prevState) => [...prevState, selectedCards[0]]);
+      setCardStates(() => [...copyOfCardStates]);
+    }
+    // if there is no match or not enough matches, flipped cards get
+    // closed by setting isFlipped to false
+    else if (NUM_REQUIRED_MATCHES === selectedCards.length) {
+      selectedCards.forEach((card) => {
+        // current index is needed to keep the order of cards displayed
+        // because goal is to make user remember the spots of previous cards
+        const targetIndex = copyOfCardStates.findIndex(
+          (cardState) => cardState.cardKey === card.cardKey
+        );
+        copyOfCardStates[targetIndex] = {
+          ...copyOfCardStates[targetIndex],
+          isFlipped: false,
+        };
+      });
+      // this gives user 750ms to see failed match by delaying setting isFlipped to false.
+      // Otherwise game gets too punishing
+      setTimeout(() => {
+        setCardStates(() => [...copyOfCardStates]);
+      }, 750);
```

Comment:
magic number, move to top, give useful name

Reply:
PTAL